File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.py
import sys
input = sys.stdin.readline

n = int(input())
egg = [list(map(int,input().split())) for _ in range(n)]
ans = 0

def bt(now,k):
    global ans
    
    if now == n:
        ans = max(ans,k)
        return
    
    if egg[now][0] <= 0:
        bt(now+1,k)
    else:
        for i in range(n):
            if i == now:
                continue
            if egg[i][0] > 0:
                egg[now][0] -= egg[i][1]
                egg[i][0] -= egg[now][1]
                if egg[now][0] <= 0:
                    k += 1
                if egg[i][0] <= 0:
                    k += 1
                bt(now+1,k)
                if egg[now][0] <= 0:
                    k -= 1
                if egg[i][0] <= 0:
                    k -= 1
                egg[now][0] += egg[i][1]
                egg[i][0] += egg[now][1]
            else:
                bt(now+1,k)
            
bt(0,0)
print(ans)

File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, ans, cnt;
	static Egg eggs[];

	static class Egg {
		int s, w;

		public Egg(int s, int w) {
			this.s = s;
			this.w = w;
		}
	}

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		n = Integer.parseInt(br.readLine());
		eggs = new Egg[n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(st.nextToken());
			int w = Integer.parseInt(st.nextToken());
			eggs[i] = new Egg(s, w);
		}
		ans = 0;
		cnt = 0;
		solve(0);
		System.out.println(ans);
	}

	static void solve(int idx) {
		if (idx == n || cnt == n) {
			ans = Math.max(ans, cnt);
			return;
		}
		if (eggs[idx].s < 1) {
			solve(idx + 1);
		} else {
			boolean flag = false;
			for (int i = 0; i < n; i++) {
				if (i == idx)
					continue;
				if (eggs[i].s > 0) {
					eggs[idx].s -= eggs[i].w;
					eggs[i].s -= eggs[idx].w;
					if (eggs[idx].s <= 0)
						cnt++;
					if (eggs[i].s <= 0)
						cnt++;
					flag = true;
					solve(idx + 1);
					if (eggs[idx].s <= 0)
						cnt--;
					if (eggs[i].s <= 0)
						cnt--;
					eggs[idx].s += eggs[i].w;
					eggs[i].s += eggs[idx].w;
				}
			}
			if (!flag)
				solve(idx + 1);
		}
	}
}